SnakeGame/Snake.js

```diff
@@ -0,0 +1,207 @@
+class Snake {
+    constructor(game) {
+        this.game = game;
+        this.direction = "R";
+        this.squeue = new Queue();
+        this.head = [2, 3];
+        // start off as a 3 unit long snake
+        this.squeue.enqueue([2, 1]);
+        this.squeue.enqueue([2, 2]);
+        this.squeue.enqueue([2, 3]);
+        this.appleCount = 0;
+        this.input = "R";
+
+        pageCounter.textContent = "Apples Eaten: " + this.appleCount;
+    }
+
+    move() {
+        let dir = this.direction;
+        if (this.input == "R" && dir == "L"
+            || this.input == "L" && dir == "R"
+            || this.input == "U" && dir == "D"
+            || this.input == "D" && dir == "U") {
+        } else {
+            this.direction = this.input;
+        }
+        dir = this.direction;
+
+        if (dir == "R") {
+            this.squeue.enqueue([this.head[0], this.head[1] + 1]);
+            this.head = [this.head[0], this.head[1] + 1];
+        } else if (dir == "U") {
+            this.squeue.enqueue([this.head[0] - 1, this.head[1]]);
+            this.head = [this.head[0] - 1, this.head[1]];
+        } else if (dir == "L") {
+            this.squeue.enqueue([this.head[0], this.head[1] - 1]);
+            this.head = [this.head[0], this.head[1] - 1];
+        } else if (dir == "D") {
+            this.squeue.enqueue([this.head[0] + 1, this.head[1]]);
+            this.head = [this.head[0] + 1, this.head[1]];
+        }
+        let tail;
+
+        // check if apple is eaten
+        let apple = game.apple;
+        if (this.head[0] == apple[0] && this.head[1] == apple[1]) {
+            tiles[apple[0]][apple[1]].classList.remove("apple");
+            this.appleCount++;
+            this.game.apple = game.newAppleLoc();
+            pageCounter.textContent = "Apples Eaten: " + this.appleCount;
+        } else {
+            tail = this.squeue.dequeue().data;
+        }
+
+        game.tick(this.head, tail);
+    }
+}
+
+class Game {
+    constructor(n) {
+        this.n = n;
+    }
+
+    reset() {
+        this.board = [];
```

Comment:
This initialization can be reduced to:
```
this.board = [...Array(this.n)].map(() => Array(this.n).fill("-"));
```
The understanding behind this is:
- `Array(this.n).fill("-")` does the role of the inner loop. This creates an array like `["-", "-", "-", ... "-"]`.
- We can't use `Array(this.n).fill(Array(this.n).fill("-"))` for the outer loop: the reason is because it would create copies of an array (which are references to the same array, since arrays are not primitive; this is problematic for us). Thus we need to create a new array for each row, which is achieved by using a `.map()` function. `[...Array(this.n)]` gives us an empty array of length `this.n`, and we turn each element into a new row.